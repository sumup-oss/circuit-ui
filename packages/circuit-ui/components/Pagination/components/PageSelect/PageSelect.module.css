.total {
  margin-left: var(--cui-spacings-kilo);
}
