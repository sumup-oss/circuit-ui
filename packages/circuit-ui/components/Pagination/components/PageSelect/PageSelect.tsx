/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useCallback, FunctionComponent, ChangeEvent, Fragment } from 'react';

import Select, { SelectProps } from '../../../Select/index.js';

import classes from './PageSelect.module.css';

export interface PageSelectProps extends Omit<SelectProps, 'onChange'> {
  onChange: (page: number) => void;
  pages: number[];
  currentPage: number;
  totalPages: number;
  totalLabel?: (totalPages: number) => string;
  [key: string]: unknown;
}

export const PageSelect: FunctionComponent<PageSelectProps> = ({
  label,
  onChange,
  pages,
  currentPage,
  totalPages,
  totalLabel,
  ...props
}) => {
  const pageOptions = pages.map((value) => ({ value, label: `${value}` }));
  const handleChange = useCallback(
    (event: ChangeEvent<HTMLSelectElement>) => {
      onChange(parseInt(event.target.value, 10));
    },
    [onChange],
  );

  return (
    <Fragment>
      <Select
        {...props}
        label={label}
        value={currentPage}
        options={pageOptions}
        onChange={handleChange}
        hideLabel
      />
      {totalLabel && (
        <span className={classes.total}>{totalLabel(totalPages)}</span>
      )}
    </Fragment>
  );
};
