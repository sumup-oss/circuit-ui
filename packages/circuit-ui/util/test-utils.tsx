/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
