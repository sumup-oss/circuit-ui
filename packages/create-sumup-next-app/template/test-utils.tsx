import React, { FunctionComponent } from 'react';
import { configureAxe } from 'jest-axe';
import { ThemeProvider } from '@emotion/react';
import { render as renderTest } from '@testing-library/react';
import { light } from '@sumup/design-tokens';

const WithProviders: FunctionComponent = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);

export const render = (
  component: Parameters<typeof renderTest>[0],
  options: Parameters<typeof renderTest>[1] = {},
) => renderTest(component, { wrapper: WithProviders, ...options });

export const axe = configureAxe({
  rules: {
    // Disable landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export * from '@testing-library/react';
