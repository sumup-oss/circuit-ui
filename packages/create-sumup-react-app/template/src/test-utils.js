import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';
import PropTypes from 'prop-types';

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={light}>{children}</ThemeProvider>
);
AllTheProviders.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';

export { customRender as render };
