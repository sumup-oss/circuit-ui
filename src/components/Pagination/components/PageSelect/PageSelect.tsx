/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, FunctionComponent, ChangeEvent } from 'react';

import styled from '../../../../styles/styled';
import { Select, SelectProps } from '../../../Select/Select';

export interface PageSelectProps extends Omit<SelectProps, 'onChange'> {
  onChange: (page: number) => void;
  pages: number[];
  currentPage: number;
  totalPages: number;
  totalLabel?: (totalPages: number) => string;
  [key: string]: any;
}

const TotalPages = styled('span')`
  margin-left: ${p => p.theme.spacings.kilo};
`;

export const PageSelect: FunctionComponent<PageSelectProps> = ({
  label,
  onChange,
  pages,
  currentPage,
  totalPages,
  totalLabel,
  ...props
}) => {
  const pageOptions = pages.map(value => ({ value, label: `${value}` }));

  const handleChange = useCallback(
    (event: ChangeEvent<HTMLSelectElement>) => {
      onChange(parseInt(event.target.value, 10));
    },
    [onChange]
  );

  return (
    <>
      <Select
        {...props}
        label={label}
        value={currentPage}
        options={pageOptions}
        onChange={handleChange}
        hideLabel
        noMargin
      />
      {totalLabel && <TotalPages>{totalLabel(totalPages)}</TotalPages>}
    </>
  );
};
