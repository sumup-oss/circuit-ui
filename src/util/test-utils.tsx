/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FunctionComponent } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, wait, act } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import ComponentsContext, {
  defaultComponents,
} from '../components/ComponentsContext';

export type RenderFn = (component: React.ReactElement, ...rest: any) => any;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const renderWithProviders = (renderer: RenderFn): RenderFn => (
  component,
  ...rest
): RenderFn => renderer(<WithProviders>{component}</WithProviders>, rest);

const render: RenderFn = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn = renderWithProviders(renderToStaticMarkup);
const create: RenderFn = (...args) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export {
  create,
  render,
  renderToHtml,
  renderHook,
  act,
  actHook,
  userEvent,
  wait,
  axe,
};
