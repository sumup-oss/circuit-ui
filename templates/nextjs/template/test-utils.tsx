import { configureAxe } from 'jest-axe';

export const axe = configureAxe({
  rules: {
    // Disable landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export * from '@testing-library/react';
