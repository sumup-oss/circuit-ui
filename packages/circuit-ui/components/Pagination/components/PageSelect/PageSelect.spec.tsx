/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it, vi } from 'vitest';

import { render, fireEvent, axe, screen } from '../../../../util/test-utils.js';

import { PageSelect, PageSelectProps } from './PageSelect.js';

describe('PageSelect', () => {
  const baseProps: PageSelectProps = {
    onChange: vi.fn(),
    label: 'Pagination',
    totalLabel: (total) => `of ${total}`,
    pages: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    currentPage: 1,
    totalPages: 10,
  };

  it('should call the onChange callback', () => {
    const testId = 'select-page';
    const onChange = vi.fn();
    render(
      <PageSelect {...baseProps} onChange={onChange} data-testid={testId} />,
    );
    const selectEl = screen.getByTestId(testId);

    fireEvent.change(selectEl, { target: { value: '3' } });

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(3);
  });

  it('should have no accessibility violations', async () => {
    const { container } = render(<PageSelect {...baseProps} />);
    const actual = await axe(container);
    expect(actual).toHaveNoViolations();
  });
});
