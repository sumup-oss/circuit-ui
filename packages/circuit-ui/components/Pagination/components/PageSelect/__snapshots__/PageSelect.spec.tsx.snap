// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSelect styles should render with default styles 1`] = `
HTMLCollection [
  .circuit-0[disabled],
.circuit-0.cui-field-disabled {
  pointer-events: none;
}

.circuit-1 {
  display: block;
  font-size: 0.875rem;
  line-height: 1.25rem;
}

.circuit-2 {
  display: inline-block;
  margin-bottom: 4px;
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

[disabled] .circuit-2,
.cui-field-disabled .circuit-2 {
  color: var(--cui-fg-normal-disabled);
}

.circuit-4 {
  color: var(--cui-fg-normal);
  display: block;
  position: relative;
}

.circuit-5 {
  font-size: 1rem;
  line-height: 1.5rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  cursor: pointer;
  background-color: var(--cui-bg-normal);
  outline: none;
  border: 0;
  border-radius: 8px;
  box-shadow: none;
  color: var(--cui-fg-normal);
  margin: 0;
  padding-top: 12px;
  padding-right: 48px;
  padding-bottom: 12px;
  padding-left: 16px;
  position: relative;
  width: 100%;
  z-index: 20;
  overflow-x: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  transition: box-shadow 120ms ease-in-out,padding 120ms ease-in-out;
  box-shadow: 0 0 0 1px var(--cui-border-normal);
}

.circuit-5:-moz-focusring {
  color: transparent;
  text-shadow: 0 0 0 #000;
}

.circuit-5::-ms-expand {
  display: none;
}

.circuit-5:hover {
  box-shadow: 0 0 0 1px var(--cui-border-normal-hovered);
}

.circuit-5:focus {
  box-shadow: 0 0 0 2px var(--cui-border-accent);
}

.circuit-5:active {
  box-shadow: 0 0 0 1px var(--cui-border-accent);
}

.circuit-6 {
  color: var(--cui-fg-subtle);
  display: block;
  z-index: 21;
  pointer-events: none;
  position: absolute;
  height: 48px;
  width: 48px;
  top: 0;
  right: 0;
  padding: 16px;
}

select:active~.circuit-6 {
  display: none;
}

.circuit-7 {
  color: var(--cui-fg-subtle);
  display: block;
  z-index: 21;
  pointer-events: none;
  position: absolute;
  height: 48px;
  width: 48px;
  top: 0;
  right: 0;
  padding: 16px;
}

select:not(:active)~.circuit-7 {
  display: none;
}

<div
    class=" circuit-0"
  >
    <label
      class="circuit-1"
      for="select_1"
    >
      <span
        class="circuit-2 circuit-3"
      >
        Pagination
      </span>
    </label>
    <div
      class="circuit-4"
    >
      <select
        aria-describedby="validation-hint_2"
        class="circuit-5"
        id="select_1"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="7"
        >
          7
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="9"
        >
          9
        </option>
        <option
          value="10"
        >
          10
        </option>
      </select>
      <svg
        aria-hidden="true"
        class="circuit-6"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 11c-.266 0-.52-.104-.71-.29l-4-4A1.013 1.013 0 0 1 3 6a1 1 0 0 1 1-1c.266 0 .52.104.71.29L8 8.58l3.29-3.29c.19-.186.445-.29.71-.29a1 1 0 0 1 1 1c0 .266-.104.52-.29.71l-4 4c-.19.186-.444.29-.71.29z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="circuit-7"
        fill="none"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 5c.266 0 .52.104.71.29l4 4c.186.19.29.444.29.71a1 1 0 0 1-1 1c-.265 0-.52-.104-.71-.29L8 7.42l-3.29 3.29c-.19.186-.444.29-.71.29a1 1 0 0 1-1-1c0-.266.104-.52.29-.71l4-4C7.48 5.104 7.734 5 8 5z"
          fill="currentColor"
        />
      </svg>
    </div>
    <span
      aria-live="polite"
      role="status"
    />
  </div>,
  .circuit-0 {
  margin-left: 12px;
}

<span
    class="circuit-0"
  >
    of 10
  </span>,
]
`;
